File: Week_09/541/reverse_string_ii.py
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        chars = list(s)
        i = 0
        n = len(s)
        while i < n:
            start = i
            end = min(i + k - 1, n - 1)
            while start < end:
                chars[start], chars[end] = chars[end], end[start]
                start += 1
                end -= 1
            i += 2 * k
        return str(chars)

File: Week_09/541/reverse_string_ii.go
package main

func reverseStr(s string, k int) string {
	bytes := []byte(s)
	n := len(bytes)
	for i:=0; i < n; {
		if n - i < k {
			reverseHelper(bytes, i, n-1)
		} else {
			reverseHelper(bytes, i, i+k-1)
		}
		i  += 2*k
	}
	return string(bytes)
}

func reverseHelper(bytes []byte, start int, end int) {
	for start < end {
		bytes[start], bytes[end] = bytes[end], bytes[start]
		start++
		end--
	}
}